changefeedccl: fix premature shutdown due to schema change bug

andyyang890 opened the pull request:
Fixes #144108

Test failures on master:
Fixes #143976
Fixes #144045
Fixes #144219

Test failures on release branches (won't be fixed until PR is backported):
Informs #144287
Informs #144291
Informs #144352

---

**changefeedccl: add more verbose logging around schema changes**

This patch adds more verbose logging to the change aggregator around
receiving and emitting resolved spans to help debug recurring changefeed
schema change test flakes.

Release note: None

---

**changefeedccl: fix premature shutdown due to schema change bug**

This patch fixes a bug that could potentially cause a changefeed
to erroneously complete when one of its watched tables encounters a
schema change has been fixed.

The root cause for the bug was that if we happened to get a rangefeed
checkpoint at precisely `ts.Prev()` for some schema change timestamp
`ts`, the kv feed would deliver a resolved span with `ts` and a NONE
boundary to the change aggregator, which would advance its frontier;
then when the resolved span with `ts` and a RESTART boundary was
sent to the change aggregator, the frontier would not be advanced and
so would not be flushed to the change frontier. The change frontier
would then read a nil row from the change aggregator and shut the
changefeed down as if it had completed successfully.

This bug has been fixed by modifying the resolved span frontier
forwarding logic to consider forwarding to the current timestamp
but with a non-NONE boundary type to be advancing the frontier.

Two alternative solutions that were ruled out were:
1. Unconditionally flushing the frontier when we get a non-NONE boundary
   instead of only flushing when the frontier is advanced.
     - Problem: We would flush the frontier O(spans) number of times.
2. Making the kv feed not emit resolved events for rangefeed checkpoints
   that are at a schema change boundary.
     - Problem: We wouldn't be able to save the per-span progress at the
       schema change boundary.

Release note (bug fix): A bug that could potentially cause a changefeed
to erroneously complete when one of its watched tables encounters a
schema change has been fixed.

---

**changefeedccl/resolvedspan: update assertion for forwarding boundary** 

This patch adds an assertion that the frontier will not forward the
boundary to a different type at the same time. This assertion is
important because if it's violated, the changefeed processors will not
shut down correctly during a schema change. One potential way this
could be violated in the future is a change to how boundary types are
determined on aggregators causing different boundaries to be sent from
aggregators in a mixed-version cluster for the same schema change.

Release note: None

andyyang890 committed:
changefeedccl: add more verbose logging around schema changes

This patch adds more verbose logging to the change aggregator around
receiving and emitting resolved spans to help debug recurring changefeed
schema change test flakes.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144004)
<!-- Reviewable:end -->

andyyang890 committed:
changefeedccl: fix premature shutdown due to schema change bug

This patch fixes a bug that could potentially cause a changefeed
to erroneously complete when one of its watched tables encounters a
schema change has been fixed.

The root cause for the bug was that if we happened to get a rangefeed
checkpoint at precisely `ts.Prev()` for some schema change timestamp
`ts`, the kv feed would deliver a resolved span with `ts` and a NONE
boundary to the change aggregator, which would advance its frontier;
then when the resolved span with `ts` and a RESTART boundary was
sent to the change aggregator, the frontier would not be advanced and
so would not be flushed to the change frontier. The change frontier
would then read a nil row from the change aggregator and shut the
changefeed down as if it had completed successfully.

This bug has been fixed by modifying the resolved span frontier
forwarding logic to consider forwarding to the current timestamp
but with a non-NONE boundary type to be advancing the frontier.

Two alternative solutions that were ruled out were:
1. Unconditionally flushing the frontier when we get a non-NONE boundary
   instead of only flushing when the frontier is advanced.
     - Problem: We would flush the frontier O(spans) number of times.
2. Making the kv feed not emit resolved events for rangefeed checkpoints
   that are at a schema change boundary.
     - Problem: We wouldn't be able to save the per-span progress at the
       schema change boundary.

Release note (bug fix): A bug that could potentially cause a changefeed
to erroneously complete when one of its watched tables encounters a
schema change has been fixed.

andyyang890 commented:
Two alternative solutions that I ruled out:
1. Unconditionally flushing the frontier when we get a non-NONE boundary here: https://github.com/cockroachdb/cockroach/blob/7ec4e46c22c432c24644a387fa583358b5fd0c70/pkg/ccl/changefeedccl/changefeed_processors.go#L900-L901
    * Problem: we would end up flushing the frontier O(spans) number of times.
2. Making the kv feed not emit resolved events when it's at the boundary of a schema change (`ts.Prev()` for some schema change timestamp `ts`).
    * Problem: we wouldn't be able to per-span checkpoint progress at this timestamp.

andyyang890 committed:
changefeedccl/resolvedspan: update assertion for forwarding boundary

This patch adds an assertion that the frontier will not forward the
boundary to a different type at the same time. This assertion is
important because if it's violated, the changefeed processors will not
shut down correctly during a schema change. One potential way this
could be violated in the future is a change to how boundary types are
determined on aggregators causing different boundaries to be sent from
aggregators in a mixed-version cluster for the same schema change.

Release note: None

rharding6373 reviewed (changes requested):
Nice! What's the backport plan for the other releases?

andyyang890 commented on pkg/ccl/changefeedccl/resolvedspan/frontier.go:
```diff
@@ -161,11 +163,13 @@ func (f *CoordinatorFrontier) All() iter.Seq[jobspb.ResolvedSpan] {
+					},
+				); ok {
+					resolvedSpan.BoundaryType = jobspb.ResolvedSpan_BACKFILL
+					break
```
Great catch, thanks

andyyang890 commented on pkg/ccl/changefeedccl/resolvedspan/frontier.go:
```diff
@@ -224,8 +224,13 @@ func newResolvedSpanFrontier(
-func (f *resolvedSpanFrontier) ForwardResolvedSpan(r jobspb.ResolvedSpan) (bool, error) {
-	forwarded, err := f.Forward(r.Span, r.Timestamp)
+// The frontier is considered forwarded if either the frontier
+// timestamp advances or the boundary is advanced and the frontier
```
Good point, I updated the terminology to hopefully be more clear now

andyyang890 commented on pkg/ccl/changefeedccl/resolvedspan/frontier.go:
```diff
@@ -235,7 +240,12 @@ func (f *resolvedSpanFrontier) ForwardResolvedSpan(r jobspb.ResolvedSpan) (bool,
+		boundaryForwarded := f.boundary.Forward(newBoundary)
+		if boundaryForwarded && !forwarded {
+			// The frontier is considered forwarded if the boundary
+			// was advanced and we are already at the boundary.
```
Updated for clarity, thanks

andyyang890 commented:
> What's the backport plan for the other releases?

We can manually backport once we're confident this fixes the issue. As discussed offline, backporting to <25.1 will require a different set of code changes because the resolved span frontier was refactored in 25.1.

rharding6373 reviewed (approved):
Nice work!

andyyang890 commented:
TFTR!

bors r=rharding6373

andyyang890 commented:
bors retry

andyyang890 commented:
bors retry
